InnerEye/ML/lightning_models.py

```diff
@@ -0,0 +1,980 @@
+                 use_average_across_structures: bool = True) -> None:
+        """
+        Creates a new MetricForMultipleStructures object.
+        :param ground_truth_ids: The list of anatomical structures that should be stored.
+        :param metric_name: The name of the metric that should be stored. This is used in the names of the individual
+        metrics.
+        :param is_training: If true, use "train/" as the prefix for all metric names, otherwise "val/"
+        :param use_average_across_structures: If True, keep track of the average metric value across structures,
+        while skipping NaNs. If false, only store the per-structure metric values.
+        """
+        super().__init__()
+        prefix = (TRAIN_PREFIX if is_training else VALIDATION_PREFIX) + metric_name + "/"
+        # All Metric classes must be
+        self.average_per_structure = ModuleList([AverageWithoutNan(name=prefix + g) for g in ground_truth_ids])
+        self.use_average_across_structures = use_average_across_structures
+        if use_average_across_structures:
+            self.average_all = AverageWithoutNan(name=prefix + AVERAGE_DICE_SUFFIX)
+        self.count = len(ground_truth_ids)
+
+    def update(self, values_per_structure: torch.Tensor) -> None:
+        """
+        Stores a vector of per-structure Dice scores in the present object. It updates the per-structure values,
+        and the aggregate value across all structures.
+        :param values_per_structure: A row tensor that has as many entries as there are ground truth IDs.
+        """
+        if values_per_structure.dim() != 1 or values_per_structure.numel() != self.count:
+            raise ValueError(f"Expected a tensor with {self.count} elements, but "
+                             f"got shape {values_per_structure.shape}")
+        for i, v in enumerate(values_per_structure.view((-1,))):
+            self.average_per_structure[i].update(v)
+        if self.use_average_across_structures:
+            self.average_all.update(nanmean(values_per_structure))
+
+    def __iter__(self) -> Iterator[Metric]:
+        """
+        Enumerates all the metrics that the present object holds: First the average across all structures,
+        then the per-structure Dice scores.
+        """
+        if self.use_average_across_structures:
+            yield self.average_all
+        yield from self.average_per_structure
+
+    def compute_all(self) -> Iterator[Tuple[str, torch.Tensor]]:
+        """
+        Calls the .compute() method on all the metrics that the present object holds, and returns a sequence
+        of (metric name, metric value) tuples. This will automatically also call .reset() on the metrics.
+        The first returned metric is the average across all structures, then come the per-structure values.
+        """
+        for d in iter(self):
+            yield d.name, d.compute()  # type: ignore
+
+
+class InnerEyeLightning(LightningModule):
```

Comment:
In our applications, we override some base methods of pl.module including:
`on_train_batch_end`, `on_validation_batch_end` etc .. The application specific changes are not necessarily only in the validation and training steps function `training_or_validation_step`. Therefore, is there any documentation on how users can bring their own of lightining module that does not inherit from inneryelightining but still be able to use the other important functionalities of the codebase.

For instance, if users finds other IE features useful but does not need timers, how could they integrate their PL modules? Would there be a need to simplify the baseclass further in terms of signatures and optimisers etc.

Reply:
Yes, that is true. I think that's a good point to keep in mind for enabling "bring your own PL model". However, at the moment, we don't officially support that yet, and we don't have a test case for it. I'll close that for now.